source/contracts/reporting/Market.sol

```diff
@@ -177,14 +176,53 @@ contract Market is DelegationTarget, Typed, Initializable, Ownable, IMarket {
         IReportingToken _tentativeWinningReportingToken = reportingTokens[tentativeWinningPayoutDistributionHash];
         if (address(_tentativeWinningReportingToken) == NULL_ADDRESS) {
             tentativeWinningPayoutDistributionHash = _payoutDistributionHash;
-            _tentativeWinningReportingToken = _reportingToken;
+            return true;
         }
-        if (_reportingToken.totalSupply() > _tentativeWinningReportingToken.totalSupply()) {
+
+        bytes32 _secondPlacePayoutDistributionHash = bestGuessSecondPlaceTentativeWinningPayoutDistributionHash;
+        int256 _tentativeWinningStake = getPayoutDistributionHashStake(_tentativeWinningReportingToken);
+        int256 _payoutStake = getPayoutDistributionHashStake(_reportingToken);
+
+        if (_payoutStake > _tentativeWinningStake) {
+            bestGuessSecondPlaceTentativeWinningPayoutDistributionHash = tentativeWinningPayoutDistributionHash;
             tentativeWinningPayoutDistributionHash = _payoutDistributionHash;
         }
+
+        if (_secondPlacePayoutDistributionHash != bytes32(0)) {
+            IReportingToken _secondPlaceWinningReportingToken = reportingTokens[_secondPlacePayoutDistributionHash];
+            int256 _secondPlaceTentativeWinningStake = getPayoutDistributionHashStake(_secondPlaceWinningReportingToken);
+            if (_secondPlaceTentativeWinningStake > _tentativeWinningStake) {
+                bestGuessSecondPlaceTentativeWinningPayoutDistributionHash = tentativeWinningPayoutDistributionHash;
+                tentativeWinningPayoutDistributionHash = _secondPlacePayoutDistributionHash;
+            } else if (_secondPlacePayoutDistributionHash > bestGuessSecondPlaceTentativeWinningPayoutDistributionHash) {
+                bestGuessSecondPlaceTentativeWinningPayoutDistributionHash = _secondPlacePayoutDistributionHash;
+            }
+        }
+
```

Comment:
This logic _feels_ more complicated than necessary.  Fundamentally, we are just trying to find the first and second place hashes given the stake behind each of the potentially 3 hashes.  If we tread non-existant hashes as 0 stake and null hashes as 0 stake, I believe the following if-else brick will come to the correct solution in a way that I _think_ is easier to read (open to feedback on the truth of that statement).  It is effectively a really naive sorting algorithm (since building a real sorting algorithm is a pain in Solidity when dealing with complex datatypes (hash + value).

```
if (stakeA > stakeB && stakeB > stakeC) {
  winningHash = stakeA;
  secondPlaceHash = stakeB;
} else if (stakeA > stakeC && stakeC > stakeB) {
  winningHash = stakeA;
  secondPlaceHash = stakeC;
} else if (stakeB > stakeA && stakeA > stakeC) {
  winningHash = stakeB;
  secondPlaceHash = stakeA;
} else if (stakeB > stakeC && stakeC > stakeA) {
  winningHash = stakeB;
  secondPlaceHash = stakeC;
} else if (stakeC > stakeA && stakeA > stakeB) {
  winningHash = stakeC;
  secondPlaceHash = stakeA;
} else if (stakeC > stakeB && stakeB > stakeA) {
  winningHash = stakeC;
  secondPlaceHash = stakeB;
}
```

Reply:
Yours has less indentation and is probably more intuitive so I'll go with it!

Reply:
Well, there is the issue that treating non existent hashes and null hashes at 0 stake will incorrectly promote them in some scenarios where we have negative values. I'll think about how we could still organize this better.

Reply:
Perhaps it is reasonable to consider anything with `0` stake to be `0` hash?  `secondPlaceHash = (stakeA != 0) ? hashA : bytes32(0);` for each of the assignment lines?

Reply:
That worries me for a case where somehow legitimate staked REP cancels out a bond to result at 0. I know it's unlikely but it'd be possible.

Would it be too inelegant to have the `getPayoutDistributionHashStake` function return the smallest possible value based on negative stake from bonds if the has or token doesn't exist? That would make the above work quite nicely I think.

Reply:
If a bond cancels out stake and brings it to 0 I think that should count as "not reported on" for the sake of punting the market to the next window.  That is, I don't think any outcome should be able to win with <= 0 stake.

Reply:
Could we have `getPayoutDistributionHashStake` return -2^255 (INT_MIN) if the payout hash doesn't exist?  Would that help simplify things in the way you suggested?  If we require at least positive stake (> 0), do we still need to have that function return a anything other than `0` for null hashes?

Reply:
If we want to enforce negative hashes being unable to win we'll need to add some tests and _may_ need to have additional code written as currently a negatively staked outcome can win (if the only two staked outcomes are negative and one is greater for example).

If we go with that rule then it should be fine to have the function return 0 for those, although as I said I _think_ we'll need to add additional code to validate the tentative winning hash as > 0 in some places

Reply:
In that case I think my vote would be to add the necessary code to disallow winning with <= 0 stake.  An outcome that has negative stake definitely feels like it shouldn't win because reporters have effectively asserted that "some outcome _other_ than this one is correct".  If we have to build out code to deal with that scenario, then extending that to include `0` seems pretty trivial and make it so we don't have to deal with the `0` boundary condition.